File: Commons/Graph/src/main/java/com/greentree/commons/graph/algorithm/lca/BruteLCAFinder.java
package com.greentree.commons.graph.algorithm.lca;

import com.greentree.commons.graph.RootTree;


public class BruteLCAFinder<V> implements LCAFinder<V> {
	
	private final RootTree<? extends V> tree;
	
	public BruteLCAFinder(RootTree<? extends V> tree) {
		this.tree = tree;
	}
	
	@SuppressWarnings("unchecked")
	@Override
	public V lca(Object a, Object b) {
		if(tree.containsInSubTree(a, b))
			return (V) b;
		return lca(a, tree.getParent(b));
	}
	
	
}

File: Commons/Graph/src/main/java/com/greentree/commons/graph/algorithm/lca/BruteLCAFinder.kt
package com.greentree.commons.graph.algorithm.lca

import com.greentree.commons.graph.RootTree

class BruteLCAFinder<out V : Any>(
	private val tree: RootTree<V>,
) : LCAFinder<V> {

	override fun lca(a: @UnsafeVariance V, b: @UnsafeVariance V): V {
		if(tree.containsInSubTree(a, b)) return b
		return lca(a, tree.getParent(b))
	}
}
